lib/AST/ASTImporter.cpp

```diff
@@ -2431,7 +2434,12 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   TypeSourceInfo *TInfo = Importer.Import(D->getTypeSourceInfo());
   if (D->getTypeSourceInfo() && !TInfo)
     return nullptr;
-
+  
+  // Check recursive import.
+  Decl *AlreadyImported = Importer.GetAlreadyImportedOrNull(D);
```

Comment:
What is the rationale to check recursive import here instead of at the beginning of the function?

Reply:
Ok, as far as I can see, we say that every operation before this point will return the already imported pointers. Does this include the parameter importing above (with the potential default argument initializer expressions)?
Also, at the beginning of this function, we try to find another function declaration with the same name and types. Should we do this work only to find out later that the function is already imported? I wonder if the performance would be better if we moved this check earlier or not.

Reply:
At the beginning of the visitor function the check would be redundant, it is checked already by the `ASTImporter::Import(Decl*)`. The check should find if the same thing was imported during the run of the visitor function by import of other things (parameters, type). This means also that the check must be after such things (import of parameter and `FromTy`) happen. At least it should be after import of `FromTy` but to be sure it was put directly before creating the new `ToFunction` object.

If the same `D` is already imported during a call of `Import` from `VisitFunctionDecl` it should be imported correctly (with every component of it: parameters, type, and those initializer expressions) because the (other) import function has been returned. But I am not sure if for every possible case a duplicate import can not make problems.

There are other places in the `ASTImporter` code where such checks are done so this can be considered as the "normal" solution for such cases (there is need for even more such checks that are revealed by the new added assertion).